File: Python/DesignPatterns/Strategy.py

# Types of trading strategies:
class RiskyTradingStrategy(object):
    def MakeTrades(self):
        print("Making risky trades!")

class ModerateTradingStrategy(object):
    def MakeTrades(self):
        print("Making moderate trades.")

class ConservativeTradingStrategy(object):
    def MakeTrades(self):
        print("Making safe trades.")

# Not Supported below Python 3.4!
from enum import Enum
class TradeConditions(Enum):
    BearMarket = 0,
    BullMarket = 1,
    RecoveringMarket = 2

# The trading firm which changes strategies based on market conditions:
class TradingFirm(object):
    def __init__(self):
        self.riskyStrategy = RiskyTradingStrategy()
        self.moderateStrategy = ModerateTradingStrategy()
        self.safeStrategy = ConservativeTradingStrategy()

        self.currentStrategy = self.moderateStrategy

    def MarketUpdte(self, tradeConditions):
        # Select the best strategy for the market conditions:
        if tradeConditions == TradeConditions.BearMarket:
            self.currentStrategy = self.safeStrategy
        elif tradeConditions == TradeConditions.BullMarket:
            self.currentStrategy = self.riskyStrategy
        elif tradeConditions == TradeConditions.RecoveringMarket:
            self.currentStrategy = self.moderateStrategy

        # Make trades with that strategy:
        self.currentStrategy.MakeTrades()

File: TypeScript/DesignPatterns/Strategy.ts
// Strategy demo in TypeScript

interface TradingStrategy
{
    makeTrades(): void;
}

class RiskyTradingStrategy implements TradingStrategy
{
    public makeTrades()
    {
        console.log("Making risky trades!");
    }
}

class ModerateTradingStrategy implements TradingStrategy
{
    public makeTrades()
    {
        console.log("Making moderate trades.");
    }
}

class ConservativeTradingStrategy implements TradingStrategy
{
    public makeTrades()
    {
        console.log("Making safe trades.");
    }
}

enum TradeConditions
{
    BearMarket,
    BullMarket, 
    RecoveringMarket
}

class TradingFirm
{
    private riskyStrategy: TradingStrategy;
    private moderateStrategy: TradingStrategy;
    private safeStrategy: TradingStrategy;

    private currentStrategy: TradingStrategy;

    constructor()
    {
        this.riskyStrategy = new RiskyTradingStrategy();
        this.moderateStrategy = new ModerateTradingStrategy();
        this.safeStrategy = new ConservativeTradingStrategy();

        this.currentStrategy = this.moderateStrategy;
    }

    public marketUpdate(marketCondition: TradeConditions)
    {
        switch(marketCondition)
        {
            case TradeConditions.BearMarket:
                this.currentStrategy = this.safeStrategy;
                break;
            case TradeConditions.BullMarket:
                this.currentStrategy = this.riskyStrategy;
                break;
            case TradeConditions.RecoveringMarket:
                this.currentStrategy = this.moderateStrategy;
                break;
        }

        this.currentStrategy.makeTrades();
    }
    
}

class StrategyDemo
{
    public execute()
    {
        let firm = new TradingFirm();

        firm.marketUpdate(TradeConditions.BullMarket); // "Making risky trades!"
        firm.marketUpdate(TradeConditions.BearMarket); // "Making safe trades."
        firm.marketUpdate(TradeConditions.RecoveringMarket); // "Making moderate trades."
    }
}